### PR TITLE
Faster BinaryArray to StringArray conversion

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #.

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: arrow-array/src/array/string_array.rs [hunk 1]
```diff
@@ -216,8 +216,19 @@ impl<OffsetSize: OffsetSizeTrait> From<GenericBinaryArray<OffsetSize>>
...
+        // We only need to validate that all values are valid UTF-8
+        let validated = std::str::from_utf8(values).unwrap();
+        for offset in offsets.iter() {
+            assert!(validated.is_char_boundary(offset.as_usize()))
+        }
+
         let builder = v.into_data().into_builder().data_type(Self::DATA_TYPE);
-        Self::from(builder.build().unwrap())
+        // SAFETY:
+        // Validated UTF-8 above
```
Comment (HaoYang670) [inline arrow-array/src/array/string_array.rs:230]: 👍
How much performance improvement can we get?

**Review by HaoYang670 — APPROVED** [review]

**Review by viirya — COMMENTED** [review]
> Faster ByteArray to StringArray conversion

You mean from BinaryArray to StringArray?

File: arrow-array/src/array/string_array.rs [hunk 2]
```diff
@@ -216,8 +216,19 @@ impl<OffsetSize: OffsetSizeTrait> From<GenericBinaryArray<OffsetSize>>
...
+        let values = v.data().buffers()[1].as_ref();
+
+        // We only need to validate that all values are valid UTF-8
+        let validated = std::str::from_utf8(values).unwrap();
+        for offset in offsets.iter() {
+            assert!(validated.is_char_boundary(offset.as_usize()))
+        }
+
         let builder = v.into_data().into_builder().data_type(Self::DATA_TYPE);
-        Self::from(builder.build().unwrap())
```
Comment (viirya) [inline arrow-array/src/array/string_array.rs:220]: Does it basically remove full validation cost?

**Review by viirya — APPROVED** [review]

File: arrow-array/src/array/string_array.rs [hunk 3]
```diff
@@ -216,8 +216,19 @@ impl<OffsetSize: OffsetSizeTrait> From<GenericBinaryArray<OffsetSize>>
...
 {
     fn from(v: GenericBinaryArray<OffsetSize>) -> Self {
+        let offsets = v.value_offsets();
+        let values = v.data().buffers()[1].as_ref();
+
+        // We only need to validate that all values are valid UTF-8
+        let validated = std::str::from_utf8(values).unwrap();
+        for offset in offsets.iter() {
+            assert!(validated.is_char_boundary(offset.as_usize()))
+        }
```
[the remark this answers is not in the document]
Comment (tustvold) [reply arrow-array/src/array/string_array.rs:226]: No, in case the string is sliced

File: arrow-array/src/array/string_array.rs (hunk 1 above)
Comment (tustvold) [reply arrow-array/src/array/string_array.rs:230]:
```
byte_array_to_string_array 20000
                        time:   [13.252 µs 13.254 µs 13.257 µs]
                        change: [-67.094% -67.073% -67.049%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 7 outliers among 100 measurements (7.00%)
  1 (1.00%) high mild
  6 (6.00%) high severe
```

**Pull request merged**